File: SBiswas/Problems/Milestone1/Problem24.cpp
#include<iostream>
#include<vector>

using namespace std;


//  Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };
 
class Solution {
public:
    TreeNode* treeBuilder(int left, int right, vector<int>& nums)
    {
        if(left>right)
        {
            return nullptr;
        }

        int mid = (left+right)/2;

        TreeNode* temp = new TreeNode(nums[mid]);
        temp->left = treeBuilder(left, mid-1, nums);
        temp->right = treeBuilder(mid+1, right, nums);

        return temp;
    }
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        int left=0, right = nums.size()-1;
        TreeNode* root = treeBuilder(left, right, nums);

        return root;
    }   
};

File: Hritwik/Problems/Milestone1/Problem24.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self):
        self.response = []
    def sortedArrayToBST(self, nums):
        if not nums:
            return None
        # Get the middle element of the sorted element to divide the array into two parts
        new_node=TreeNode(nums[len(nums)//2]) 
        new_node.left=self.sortedArrayToBST(nums[:len(nums)//2])
        new_node.right=self.sortedArrayToBST(nums[len(nums)//2+1:])
        return new_node
    def inOrderTraversal(self, root):
        def traverse(node):
            if root is None:
                return None
            if node.left:
                traverse(node.left)
            self.response.append(node.val)
            if node.right:
                traverse(node.right)
        traverse(root)
        return self.response

solution = Solution()
nums = [-10,-3,0,5,9]
response_head = solution.sortedArrayToBST(nums)
response = solution.inOrderTraversal(response_head)
print(f'InOrder Traversal of Height-balanced binary search tree of the given array: {response}')
